sdk/iot/Azure.Iot.Hub.Service/src/Authentication/SasTokenAuthenticationPolicy.cs

```diff
@@ -14,28 +15,52 @@ namespace Azure.Iot.Hub.Service.Authentication
     /// </summary>
     internal class SasTokenAuthenticationPolicy : HttpPipelinePolicy
     {
-        private readonly ISasTokenProvider _sasTokenProvider;
+        private readonly TimeSpan _getTokenTimeout = TimeSpan.FromSeconds(30);
+        private readonly IotHubSasCredential _iotHubSasCredential;
 
-        internal SasTokenAuthenticationPolicy(ISasTokenProvider sasTokenProvider)
+        internal SasTokenAuthenticationPolicy(IotHubSasCredential iotHubSasCredential)
         {
-            _sasTokenProvider = sasTokenProvider;
+            _iotHubSasCredential = iotHubSasCredential;
         }
 
         public override void Process(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline)
         {
-            AddHeaders(message);
-            ProcessNext(message, pipeline);
+            ProcessAsync(message, pipeline, false).EnsureCompleted();
         }
 
-        public override async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline)
+        public override ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline)
         {
-            AddHeaders(message);
-            await ProcessNextAsync(message, pipeline).ConfigureAwait(false);
+            return ProcessAsync(message, pipeline, true);
         }
 
-        private void AddHeaders(HttpMessage message)
+        private async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
         {
-            message.Request.Headers.Add(HttpHeader.Names.Authorization, _sasTokenProvider.GetSasToken());
+            if (message.Request.Uri.Scheme != Uri.UriSchemeHttps)
+            {
+                throw new InvalidOperationException("Bearer token authentication is not permitted for non TLS protected (https) endpoints.");
+            }
+
+            await AddHeaders(message, async).ConfigureAwait(false);
+
+            if (async)
+            {
+                await ProcessNextAsync(message, pipeline).ConfigureAwait(false);
+            }
+            else
+            {
+                ProcessNext(message, pipeline);
+            }
+        }
+
+        private async Task AddHeaders(HttpMessage message, bool async)
+        {
+            var cts = new CancellationTokenSource(_getTokenTimeout);
```

Comment:
This cancellation token is used to ensure that we don't wait for > 30secs to retrieve the sas token